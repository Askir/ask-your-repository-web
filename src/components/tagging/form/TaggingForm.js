// @flow
import React, { Component } from 'react';
import TagSelector from './TagSelector';
import TagSuggestions from './TagSuggestions';
import UploadContext from '../../upload/context/UploadContext';
import type { Image } from '../../../models/Image';
import type { Tag } from '../../../models/Tag';
import './TaggingForm.scss';

type Props = {
  image: Image,
};

class TaggingForm extends Component<Props> {
  static contextType = UploadContext;

  addTag = (tag: Tag) => {
    this.context.addTag(this.props.image.id, tag);
  };

  removeTag = (tag: Tag) => {
    this.context.removeTag(this.props.image.id, tag);
  };

  removeMultiTag = (tag: Tag) => {
    this.context.removeMultiTag(tag, this.props.image.id);
  };

  render() {
    const { image } = this.props;

    return (
      <div className="TaggingForm">
        <div className="TaggingForm__title">
          Tag your image!
        </div>
        <div className="TaggingForm__info">
          Suggestions - Type number to add
        </div>
        <TagSuggestions
          tags={image.tags}
          addTag={this.addTag}
        />
        <div className="TaggingForm__info">
          Type in a tag - Hit enter &#9166; - Repeat
        </div>
        <TagSelector
          tags={image.tags}
          addTag={this.addTag}
          removeTag={this.removeTag}
          multiTags={this.context.multiTags}
          addMultiTag={this.context.addMultiTag}
          removeMultiTag={this.removeMultiTag}
        />
        <div className="TaggingForm__tip">
          <div className="TaggingForm__tip__title">
            Multi-Tagging
          </div>
          You can press a tag to apply the tag
          <br />
          to all your uploaded images at once.
          <br />
          Multi tags will be highlighted with an extra border.
        </div>
      </div>
    );
  }
}

export default TaggingForm;
