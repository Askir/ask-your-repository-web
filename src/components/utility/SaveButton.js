// @flow
import React from 'react';
import type { Node } from 'react';
import classNames from 'classnames';
import Button from './Button';

type Props = {
  children: Node,
  className?: string,
};

function SaveButton({
  children,
  className,
  ...rest
}: Props) {
  const styleClasses = classNames('Button--save', className);

  return (
    <Button className={styleClasses} {...rest}>
      {children}
    </Button>
  );
}

SaveButton.defaultProps = {
  className: '',
};

export default SaveButton;
