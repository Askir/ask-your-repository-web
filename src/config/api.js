import axios from 'axios';
import qs from 'qs';

const api = axios.create({
  baseURL: process.env.REACT_APP_API_URL,
  requestInterceptor: config => ({ ...config }),
  paramsSerializer: parameters => qs.stringify(parameters, { arrayFormat: 'repeat' }),
});

export default api;
