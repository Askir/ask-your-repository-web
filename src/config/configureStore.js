// @flow
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../state/rootReducer';

function configureStore() {
  let middleware = [thunk];

  if (process.env.NODE_ENV !== 'production') {
    const { logger } = require('redux-logger');
    middleware = [...middleware, logger];
  }

  return createStore(
    rootReducer,
    applyMiddleware(...middleware),
  );
}

export default configureStore;
