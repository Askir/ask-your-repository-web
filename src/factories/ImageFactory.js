// @flow
import _ from 'lodash';
import faker from 'faker';
import TagFactory from './TagFactory';
import type { Image } from '../models/Image';

class ImageFactory {
  static createDummyImage(tagAmount: number = 3): Image {
    return {
      id: faker.random.number(),
      src: faker.image.dataUri(),
      tags: _.times(tagAmount, () => TagFactory.createDummyTag()),
    };
  }
}

export default ImageFactory;
