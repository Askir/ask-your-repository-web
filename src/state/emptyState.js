// @flow
import type { AppState } from './AppState';

const emptyState: AppState = {
  image: null,
  images: [],
};

export default emptyState;
