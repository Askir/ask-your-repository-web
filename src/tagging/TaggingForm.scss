@import '../style/breakpoints.scss';

.TaggingForm {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  height: 100%;
  width: 100%;
  padding: 20px;

  &__title {
    font-weight: bold;
    font-size: 1.6em;
    margin-bottom: 12px;
    color: rgba(255, 255, 255, .9);
  }

  &__info {
    color: rgba(255, 255, 255, .3);
  }

  &__button {
    margin-top: auto;
  }

  @media (min-width: $xs-to-sm) {
    min-width: 340px;
    width: 400px;
  }
}
