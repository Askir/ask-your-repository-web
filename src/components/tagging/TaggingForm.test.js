// @flow
import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import TaggingForm from './TaggingForm';
import api from '../../config/api';
import ImageFactory from '../../factories/ImageFactory';
import SaveButton from '../utility/SaveButton';
import emptyState from '../../state/emptyState';
import type { AppState } from '../../state/AppState';

const mockStore = configureStore();

jest.mock('../../config/api');

describe('<TaggingForm />', () => {
  const image = ImageFactory.createDummyImage();
  const initialState: AppState = { ...emptyState, image };

  let store;
  let wrapper;

  beforeAll(() => {
    api.post.mockImplementation(() => Promise.resolve());
  });

  beforeEach(() => {
    store = mockStore(initialState);
    wrapper = shallow(<TaggingForm store={store} />).dive();
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('form is submitted', () => {
    beforeEach(() => {
      const saveButton = wrapper.find(SaveButton);
      saveButton.simulate('click');
    });

    it('should send a POST request with the tags to /images/:id/tags', () => {
      expect(api.post).toHaveBeenCalledWith(
        `/images/${image.id}/tags`,
        { tags: image.tags },
      );
    });
  });
});
