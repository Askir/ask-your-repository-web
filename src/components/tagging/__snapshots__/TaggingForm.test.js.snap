// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaggingForm /> renders correctly 1`] = `
<div
  className="TaggingForm"
>
  <div
    className="TaggingForm__title"
  >
    Tag your image!
  </div>
  <div
    className="TaggingForm__info"
  >
    Suggestions - Type number to add
  </div>
  <Connect(TagSuggestions) />
  <div
    className="TaggingForm__info"
  >
    Type in a tag - Hit enter ⏎ - Repeat
  </div>
  <Connect(TagSelector) />
  <SaveButton
    className="TaggingForm__button"
    onClick={[Function]}
  >
    Save
  </SaveButton>
</div>
`;
